src/mumble/API.h

```diff
@@ -0,0 +1,188 @@
+													  api_promise_t *promise);
+	void requestLocalMute_v_1_0_x(mumble_plugin_id_t callerID, mumble_connection_t connection, mumble_userid_t userID,
+								  bool muted, api_promise_t *promise);
+	void requestLocalUserMute_v_1_0_x(mumble_plugin_id_t callerID, bool muted, api_promise_t *promise);
+	void requestLocalUserDeaf_v_1_0_x(mumble_plugin_id_t callerID, bool deafened, api_promise_t *promise);
+	void requestSetLocalUserComment_v_1_0_x(mumble_plugin_id_t callerID, mumble_connection_t connection,
+											const char *comment, api_promise_t *promise);
+	void findUserByName_v_1_0_x(mumble_plugin_id_t callerID, mumble_connection_t connection, const char *userName,
+								mumble_userid_t *userID, api_promise_t *promise);
+	void findChannelByName_v_1_0_x(mumble_plugin_id_t callerID, mumble_connection_t connection, const char *channelName,
+								   mumble_channelid_t *channelID, api_promise_t *promise);
+	void getMumbleSetting_bool_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, bool *outValue,
+									   api_promise_t *promise);
+	void getMumbleSetting_int_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, int *outValue,
+									  api_promise_t *promise);
+	void getMumbleSetting_double_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, double *outValue,
+										 api_promise_t *promise);
+	void getMumbleSetting_string_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, const char **outValue,
+										 api_promise_t *promise);
+	void setMumbleSetting_bool_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, bool value,
+									   api_promise_t *promise);
+	void setMumbleSetting_int_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, int value,
+									  api_promise_t *promise);
+	void setMumbleSetting_double_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, double value,
+										 api_promise_t *promise);
+	void setMumbleSetting_string_v_1_0_x(mumble_plugin_id_t callerID, mumble_settings_key_t key, const char *value,
+										 api_promise_t *promise);
+	void sendData_v_1_0_x(mumble_plugin_id_t callerID, mumble_connection_t connection, const mumble_userid_t *users,
+						  size_t userCount, const uint8_t *data, size_t dataLength, const char *dataID,
+						  api_promise_t *promise);
+	void log_v_1_0_x(mumble_plugin_id_t callerID, const char *message, api_promise_t *promise);
+	void playSample_v_1_0_x(mumble_plugin_id_t callerID, const char *samplePath, api_promise_t *promise);
+
+
+private:
+	MumbleAPI();
+
+	MumbleAPICurator m_curator;
+};
+
+/// @returns The Mumble API struct (v1.0.x)
+MumbleAPI_v_1_0_x getMumbleAPI_v_1_0_x();
+
+/// Converts from the Qt key-encoding to the API's key encoding.
+///
+/// @param keyCode The Qt key-code that shall be converted
+/// @returns The converted key code or KC_INVALID if conversion failed
+mumble_keycode_t qtKeyCodeToAPIKeyCode(unsigned int keyCode);
```

Comment:
What's the purpose of this?

Reply:
It makes the plugin API Qt-independent by using custom key codes. And this function translates Qt key codes to plugin-internal key codes.

Besides allowing us to ditch Qt if we feel like it, it also prevents plugin devs from having to pull Qt code into their plugin (which would be a huge overhead for most plugins) just to read some key codes.

Reply:
How are keycodes transmitted to plugins though?

Reply:
They get passed to the `mumble_onKeyEvent` callback. Or do you mean something else?

Reply:
Thanks, that's what I thought. I think we should have a single key list internally and externally (plugins API).

Reply:
As long as that list is not Qt's I'm fine with that. As I wrote above: Having plugin authors to pull in Qt is absolutely not an option.
Besides: As it stands I think the keycodes used for the plugin API and the ones used internally never come in contact or conflict with one another, so I currently don't see the need for them to be the same. Ultimately it'd be better of course, but as it stands I guess we'll keep using Qt codes internally and this is nothing that we can expect plugins to parse for themselves.

Reply:
My idea would be to ultimately pass `GlobalShortcuts` events to plugins.

Reply:
Note that the current API is explicitly not about global key events. Plugins only receive keyboard events as long as Mumble has focus.
If we add the possibility to set global shortcuts in the future, I think this should be done by an API function that lets the plugin register a shortcut and a callback function for when that shortcut is activated and then Mumble takes care of the rest.

In such a scenario it wouldn't be that horrible to have custom key-codes for plugins either.

Reply:
Indeed.